### PR TITLE
osd: pg merge fixes

### PR DESCRIPTION
**Pull request by liewegas**
The first patch addresses past_intervals for fabricate dmerge targets.  FIxes http://tracker.ceph.com/issues/37511.

The second patch ensures we don't get new upmaps racing with PGs that are merging.  The mgr cocntroller already clears out pg_upmaps before proposing a merge, so the new safety check just deals with the race cases.

### CODE REVIEW

**Comment by tchaikov** [comment]
```
[ RUN      ] OSDMapTest.CleanPGUpmaps
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/test/osd/TestOSDMap.cc:688: Failure
Value of: tmpmap.have_pg_upmaps(ec_pgid)
  Actual: false
Expected: true
2018-12-20 11:57:16.576 ffffa0043000  0 maybe_remove_pg_upmaps pg 1.0 is gone or merge source
[  FAILED  ] OSDMapTest.CleanPGUpmaps (14 ms)

```

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -1755,8 +1755,18 @@ void OSDMap::maybe_remove_pg_upmaps(CephContext *cct,
     to_check.insert(p.first);
   }
   for (auto& pg : to_check) {
-    if (!nextmap.pg_exists(pg)) {
-      ldout(cct, 0) << __func__ << " pg " << pg << " is gone" << dendl;
+    const pg_pool_t *pi = nextmap.get_pg_pool(pg.pool());
+    if (!pi || pg.ps() < pi->get_pg_num_pending()) {
```
Comment (xiexingguo) [inline src/osd/OSDMap.cc]: pg.ps() >= pi->get_pg_num_pending()

Comment (liewegas) [reply src/osd/OSDMap.cc]: sigh.. yes, thanks!

**Comment by xiexingguo** [comment]
@liewegas Why don't you use the "pg_pool_t::is_pending_merge" wrapper directly? Wouldn't it be simpler and more natural way for the pg-merge check?

**Comment by liewegas** [comment]
@xiexingguo yes.. i looked for that helper but not hard enough, it seems!

**Comment by tchaikov** [comment]
retest this please

**Pull request merged**